### ISSUE TITLE
False-positive errors while using typevars in overloaded functions

### ISSUE THREAD

**Issue by FasterSpeeding** [microsoft/pyright#2268]
**Describe the bug**
Pyright errors on a reportGeneralTypeIssue when typevars are included in different parts of a function's overloaded signatures due to it having been overloaded in a style which allows it to be called before being used as a decorator or directly used as a decorator

**To Reproduce**
Lint the example code provided below and watch it give the following error
```py
  test.py:22:5 - error: Overloaded function implementation is not consistent with signature of overload 2
    Type "(bar: FooT@foo | None = None, /, *, a: int = 42, b: int = 83) -> (FooT@foo | (_p0: FooT@foo) -> FooT@foo)" cannot be assigned to type "(bar: FooT@foo, /) -> FooT@foo"
      Parameter 1: type "FooT@foo" cannot be assigned to type "FooT@foo | None" (reportGeneralTypeIssues)
```

**Expected behavior**
Pyright should allow this signature where a typevar is present in different parts of it between overloads

**Screenshots or Code**
```python
import abc
import typing


class Foo(abc.ABC):
    ...


FooT = typing.TypeVar("FooT", bound=Foo)


# This signature allows for passing keyword arguments during a decorator call (e.g. `@foo(a=54)`)
@typing.overload
def foo(*, a: int = 42, b: int = 83) -> typing.Callable[[FooT], FooT]:
    ...


# This signature allows for directly doing `@foo`
@typing.overload
def foo(bar: FooT, /) -> FooT:
    ...


def foo(
    bar: typing.Optional[FooT] = None, /, *, a: int = 42, b: int = 83
) -> typing.Union[FooT, typing.Callable[[FooT], FooT]]:
    raise NotImplementedError

```

**VS Code extension or command-line**
Pyright commandline version 1.1.166

**Additional context**
I only started getting this error after 1.1.166 was released

**Comment by erictraut** [comment]
Thanks for the bug report. I made some pretty significant changes in version 1.1.166, and I introduced a small regression in this code path. Apologies for the inconvenience. I've fixed the problem and added a new test case inspired by your sample above.

**Comment by erictraut** [comment]
This is addressed in pyright 1.1.167, which I just published. It was already included in the last release of pylance.
